### ISSUE TITLE
Alternative way to verify expected calls

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#1411]
(Inspired by [this StackOverflow question](https://stackoverflow.com/questions/51904257/fakeiteasy-not-verifying-call-for-full-framework-signalr-test))

For users coming from Moq, FakeItEasy's approach to verifying calls isn't always intuitive. In Moq, you typically specify that a call is verifiable when you configure it, then call `mock.Verify()` at the end of the test:

```csharp
var mock = new Mock<IFoo>();
mock.Setup(foo => foo.Bar()).Returns(42).Verifiable();
mock.Object.Bar();
mock.Verify(); // Will fail if the call to Bar() is removed
```

FakeItEasy could support something similar:

```csharp
var foo = A.Fake<IFoo>();
A.CallTo(() => foo.Bar()).Returns(42).MustHappen();
foo.Bar();
Expected.CallsTo(foo).MustHaveHappened();
```

In addition to making migration easier for Moq users, I think it's an interesting alternative approach, because you don't need to specify the call again, and you only need one assertion per fake even if you need to assert multiple calls. It can make the tests more concise and more maintainable.

**Comment by thomaslevesque** [comment]
Comparison of the existing and suggested approaches:

- Repeat call specifications:
    ```csharp
    // Configure fake
    var foo = A.Fake<IFoo>();
    A.CallTo(() => foo.Bar(1)).Returns(1);
    A.CallTo(() => foo.Baz("hello")).Returns(true).Once();
    A.CallTo(() => foo.Frob(false)).Returns(42).Once().Then.Returns(0).Once();

    // Use the fake
    // ...

    // Verify calls
    A.CallTo(() => foo.Bar(1)).MustHaveHappened();
    A.CallTo(() => foo.Baz("hello")).MustHaveHappenedOnceExactly();
    A.CallTo(() => foo.Frob(false)).MustHaveHappenedTwiceExactly();
    ```

- Keep call specifications as variables:
    ```csharp
    // Configure fake
    var foo = A.Fake<IFoo>();
    var callToBar = A.CallTo(() => foo.Bar(1));
    callToBar.Returns(1);
    var callToBaz = A.CallTo(() => foo.Baz("hello"));
    callToBaz.Returns(true).Once();
    var callToFrob = A.CallTo(() => foo.Frob(false));
    callToFrob.Returns(42).Once().Then.Returns(0).Once();

    // Use the fake
    // ...

    // Verify calls
    callToBar.MustHaveHappened();
    callToBaz.MustHaveHappenedOnceExactly();
    callToFrob.MustHaveHappenedTwiceExactly();
    ```

- Mark calls as verifiable (Moq-style):
    ```csharp
    // Configure fake
    var foo = A.Fake<IFoo>();
    A.CallTo(() => foo.Bar(1)).Returns(1).MustHappen();
    A.CallTo(() => foo.Baz("hello")).Returns(true).Once().MustHappen();
    A.CallTo(() => foo.Frob(false)).Returns(42).Once().Then.Returns(0).Once().MustHappen();

    // Use the fake
    // ...

    // Verify calls
    Expected.CallsTo(foo).MustHaveHappened();
    ```

Pros and cons of the last approach:
- PRO: concise
- PRO: no repetition of the call specifications
- PRO: a single call to assert all expected calls
- CON: it's a different way of doing the same thing
- CON: no support for ordered assertions (might be doable, though, e.g. with a `MustHappenInOrder()` method)
- CON: might be hard to get it right with regard to the number of repetitions (e.g. should we assert for two calls if the behavior was configured `Twice()`? Should the number of repetitions be specified explicitly in `MustHappen()`? etc.)

**Comment by blairconrad** [comment]
Comprehensive summary!

As I hinted in the StackOverflow question, I'm not a fan of the feature. I've used its equivalent in NMock2 back in the day, and found that my fellows (and probably me, but an unreliable memory is a wonderful balm) ended up writing brittle tests, ending up asserting on calls that didn't need it. Whenever I can, I advise people not to assert on the calls to the fake, if there's any way to evaluate success of a test without doing so.

Of course, with your proposal, one has to explicitly enrol in the verification, so that may not be a big problem. I think with NMock2, one could accidentally fall into the "auto-verify" behaviour.

However, that's just a personal objection. Nobody _has_ to use the feature.

Putting on my package maintainer hat:

1. having the different way of doing things (as you point out) might confuse users, and could lead to more work for the maintainers
2. achieving feature parity with the existing assertions might be quite tricky. Maybe it's okay to have the feature gap, or maybe it would further confuse users. Especially the difficulties with the number of repetitions, like you said.

To sum up, I don't have any interest in the feature, and wouldn't implement it myself, but don't have a strong reason for FakeItEasy not to provide it. But if we're going to go that route, I'd recommend presenting it as an alternative flow. Which I guess is how it'd appear, hanging off the `Expected` class (unless you end up going with a different syntax).

**Comment by thomaslevesque** [comment]
Thanks for your feedback @blairconrad.
On one hand, I find this approach quite compelling because of the reduced verbosity, and I think I'd probably use it if it were implemented. On the other hand, I'm a bit afraid to go down this rabbit hole; making it consistent with the existing approach could easily become a nightmare... Maybe I'll experiment with the idea and see how it goes.

**Comment by thomaslevesque** [comment]
Interestingly, it can be implementing without touching FakeItEasy's code. Here's a quick experiment: https://gist.github.com/thomaslevesque/58d7719ebfe57f8fc11f037c7173c943

I'm not very happy with this solution, though, because it relies on ambient state, which is often a source of problems.

**Comment by blairconrad** [comment]
Nice gist. I'd expected that some level of functionality would be achievable from the outside. You had better results than I expected, though, given the amount of effort.

> ambient state, which is often a source of problems

truer words…

**Comment by adamralph** [comment]
I'm not sure I like this proposal. I'm a fan of Arrange/Act/Assert or Given/When/Then style layouts for tests. This seems to mix arrangement and assertion.

**Comment by thomaslevesque** [comment]
Closing this, as there seems to be little interest for this feature.
